<!DOCTYPE html>
<html>
<head>
<title>README.md</title>
<meta http-equiv="Content-type" content="text/html;charset=UTF-8">

<style>
/* https://github.com/microsoft/vscode/blob/master/extensions/markdown-language-features/media/markdown.css */
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

body {
	font-family: var(--vscode-markdown-font-family, -apple-system, BlinkMacSystemFont, "Segoe WPC", "Segoe UI", "Ubuntu", "Droid Sans", sans-serif);
	font-size: var(--vscode-markdown-font-size, 14px);
	padding: 0 26px;
	line-height: var(--vscode-markdown-line-height, 22px);
	word-wrap: break-word;
}

#code-csp-warning {
	position: fixed;
	top: 0;
	right: 0;
	color: white;
	margin: 16px;
	text-align: center;
	font-size: 12px;
	font-family: sans-serif;
	background-color:#444444;
	cursor: pointer;
	padding: 6px;
	box-shadow: 1px 1px 1px rgba(0,0,0,.25);
}

#code-csp-warning:hover {
	text-decoration: none;
	background-color:#007acc;
	box-shadow: 2px 2px 2px rgba(0,0,0,.25);
}

body.scrollBeyondLastLine {
	margin-bottom: calc(100vh - 22px);
}

body.showEditorSelection .code-line {
	position: relative;
}

body.showEditorSelection .code-active-line:before,
body.showEditorSelection .code-line:hover:before {
	content: "";
	display: block;
	position: absolute;
	top: 0;
	left: -12px;
	height: 100%;
}

body.showEditorSelection li.code-active-line:before,
body.showEditorSelection li.code-line:hover:before {
	left: -30px;
}

.vscode-light.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(0, 0, 0, 0.15);
}

.vscode-light.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(0, 0, 0, 0.40);
}

.vscode-light.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

.vscode-dark.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(255, 255, 255, 0.4);
}

.vscode-dark.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(255, 255, 255, 0.60);
}

.vscode-dark.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

.vscode-high-contrast.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(255, 160, 0, 0.7);
}

.vscode-high-contrast.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(255, 160, 0, 1);
}

.vscode-high-contrast.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

img {
	max-width: 100%;
	max-height: 100%;
}

a {
	text-decoration: none;
}

a:hover {
	text-decoration: underline;
}

a:focus,
input:focus,
select:focus,
textarea:focus {
	outline: 1px solid -webkit-focus-ring-color;
	outline-offset: -1px;
}

hr {
	border: 0;
	height: 2px;
	border-bottom: 2px solid;
}

h1 {
	padding-bottom: 0.3em;
	line-height: 1.2;
	border-bottom-width: 1px;
	border-bottom-style: solid;
}

h1, h2, h3 {
	font-weight: normal;
}

table {
	border-collapse: collapse;
}

table > thead > tr > th {
	text-align: left;
	border-bottom: 1px solid;
}

table > thead > tr > th,
table > thead > tr > td,
table > tbody > tr > th,
table > tbody > tr > td {
	padding: 5px 10px;
}

table > tbody > tr + tr > td {
	border-top: 1px solid;
}

blockquote {
	margin: 0 7px 0 5px;
	padding: 0 16px 0 10px;
	border-left-width: 5px;
	border-left-style: solid;
}

code {
	font-family: Menlo, Monaco, Consolas, "Droid Sans Mono", "Courier New", monospace, "Droid Sans Fallback";
	font-size: 1em;
	line-height: 1.357em;
}

body.wordWrap pre {
	white-space: pre-wrap;
}

pre:not(.hljs),
pre.hljs code > div {
	padding: 16px;
	border-radius: 3px;
	overflow: auto;
}

pre code {
	color: var(--vscode-editor-foreground);
	tab-size: 4;
}

/** Theming */

.vscode-light pre {
	background-color: rgba(220, 220, 220, 0.4);
}

.vscode-dark pre {
	background-color: rgba(10, 10, 10, 0.4);
}

.vscode-high-contrast pre {
	background-color: rgb(0, 0, 0);
}

.vscode-high-contrast h1 {
	border-color: rgb(0, 0, 0);
}

.vscode-light table > thead > tr > th {
	border-color: rgba(0, 0, 0, 0.69);
}

.vscode-dark table > thead > tr > th {
	border-color: rgba(255, 255, 255, 0.69);
}

.vscode-light h1,
.vscode-light hr,
.vscode-light table > tbody > tr + tr > td {
	border-color: rgba(0, 0, 0, 0.18);
}

.vscode-dark h1,
.vscode-dark hr,
.vscode-dark table > tbody > tr + tr > td {
	border-color: rgba(255, 255, 255, 0.18);
}

</style>

<style>
/* Tomorrow Theme */
/* http://jmblog.github.com/color-themes-for-google-code-highlightjs */
/* Original theme - https://github.com/chriskempson/tomorrow-theme */

/* Tomorrow Comment */
.hljs-comment,
.hljs-quote {
	color: #8e908c;
}

/* Tomorrow Red */
.hljs-variable,
.hljs-template-variable,
.hljs-tag,
.hljs-name,
.hljs-selector-id,
.hljs-selector-class,
.hljs-regexp,
.hljs-deletion {
	color: #c82829;
}

/* Tomorrow Orange */
.hljs-number,
.hljs-built_in,
.hljs-builtin-name,
.hljs-literal,
.hljs-type,
.hljs-params,
.hljs-meta,
.hljs-link {
	color: #f5871f;
}

/* Tomorrow Yellow */
.hljs-attribute {
	color: #eab700;
}

/* Tomorrow Green */
.hljs-string,
.hljs-symbol,
.hljs-bullet,
.hljs-addition {
	color: #718c00;
}

/* Tomorrow Blue */
.hljs-title,
.hljs-section {
	color: #4271ae;
}

/* Tomorrow Purple */
.hljs-keyword,
.hljs-selector-tag {
	color: #8959a8;
}

.hljs {
	display: block;
	overflow-x: auto;
	color: #4d4d4c;
	padding: 0.5em;
}

.hljs-emphasis {
	font-style: italic;
}

.hljs-strong {
	font-weight: bold;
}
</style>

<style>
/*
 * Markdown PDF CSS
 */

 body {
	font-family: -apple-system, BlinkMacSystemFont, "Segoe WPC", "Segoe UI", "Ubuntu", "Droid Sans", sans-serif, "Meiryo";
	padding: 0 12px;
}

pre {
	background-color: #f8f8f8;
	border: 1px solid #cccccc;
	border-radius: 3px;
	overflow-x: auto;
	white-space: pre-wrap;
	overflow-wrap: break-word;
}

pre:not(.hljs) {
	padding: 23px;
	line-height: 19px;
}

blockquote {
	background: rgba(127, 127, 127, 0.1);
	border-color: rgba(0, 122, 204, 0.5);
}

.emoji {
	height: 1.4em;
}

code {
	font-size: 14px;
	line-height: 19px;
}

/* for inline code */
:not(pre):not(.hljs) > code {
	color: #C9AE75; /* Change the old color so it seems less like an error */
	font-size: inherit;
}

/* Page Break : use <div class="page"/> to insert page break
-------------------------------------------------------- */
.page {
	page-break-after: always;
}

</style>

<script src="https://unpkg.com/mermaid/dist/mermaid.min.js"></script>
</head>
<body>
  <script>
    mermaid.initialize({
      startOnLoad: true,
      theme: document.body.classList.contains('vscode-dark') || document.body.classList.contains('vscode-high-contrast')
          ? 'dark'
          : 'default'
    });
  </script>
<p><em>CPSC 290: Directed Research in Computer Science, Spring 2023, Yale University</em></p>
<h1 id="rna-discovery--analysis-at-the-breaker-lab">RNA Discovery &amp; Analysis at the Breaker Lab</h1>
<p>By Julian Sanker, advised by Dr. Smita Krishnaswamy, under the mentorship of Dr. Ronald Breaker and Christopher King.</p>
<h2 id="table-of-contents">Table of Contents</h2>
<ul>
<li><a href="#rna-discovery--analysis-at-the-breaker-lab">RNA Discovery &amp; Analysis at the Breaker Lab</a>
<ul>
<li><a href="#table-of-contents">Table of Contents</a></li>
<li><a href="#introduction">Introduction</a></li>
<li><a href="#project-components">Project Components</a>
<ul>
<li><a href="#1-backend-pipeline">1. Backend Pipeline</a></li>
<li><a href="#2-genetic-algorithm">2. Genetic Algorithm</a></li>
<li><a href="#3-frontend-website">3. Frontend Website</a></li>
</ul>
</li>
<li><a href="#methodology">Methodology</a>
<ul>
<li><a href="#code-structure">Code Structure</a></li>
<li><a href="#database-architecture">Database Architecture</a></li>
</ul>
</li>
<li><a href="#experimental-process-for-sodium-binding-riboswitches">Experimental Process for Sodium-Binding Riboswitches</a>
<ul>
<li><a href="#1-download-motifs">1. Download motifs</a></li>
<li><a href="#2-search-against-genome-database">2. Search against genome database</a></li>
<li><a href="#3-refold-using-cmfinder">3. Refold using CMfinder</a></li>
<li><a href="#4-search-new-models-against-genome-database">4. Search new models against genome database</a></li>
<li><a href="#5-evaluate-search-results">5. Evaluate search results</a></li>
<li><a href="#6-aside-report-tandem-hits">6. Aside: report tandem hits</a></li>
</ul>
</li>
<li><a href="#conclusion">Conclusion</a>
<ul>
<li><a href="#discussion">Discussion</a></li>
<li><a href="#future-research">Future Research</a></li>
<li><a href="#acknowledgements">Acknowledgements</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<h2 id="introduction">Introduction</h2>
<p><em>Project developed during Spring 2023 for CPSC 290: Directed Research in Computer Science at Yale University, under the mentorship of Dr. Ronald Breaker and Christopher King.</em></p>
<p>The main goal of the project is to improve computational tools for structured RNA discovery and analysis. The project is divided into three parts: the first is the development of a pipeline that interfaces with numerous bioinformatic software packages, including Infernal, HMMER, CMfinder, and R2R, to perform useful operations on RNA sequence data across these tools. The second is the incorporation of a genetic algorithm into the pipeline, in place of the existing, outdated CMfinder software, to predict RNA secondary structure/folding. The final component is a frontend website that provides an interface for RNA researchers to utilize the pipeline and its various operations independently, as well as a new RALEE-esque RNA alignment editor, to conduct experiments on sequence data without needing technical knowledge of the particular software packages.</p>
<p>The project is incomplete, to be continued over the summer. The following sections describe the project in more detail. The project proposal from the beginning of the semester can be found in the <a href="files/Spring%20'23%20Research%20Proposal%20CPSC%20290.pdf">link to Spring 2023 Project Proposal</a>.</p>
<h2 id="project-components">Project Components</h2>
<p>The project is divided into three components: a backend pipeline, a genetic algorithm, and a frontend website. The backend pipeline is currently being redesigned, and the genetic algorithm and frontend website are currently incomplete. The following sections describe each component in more detail.</p>
<h3 id="1-backend-pipeline">1. Backend Pipeline</h3>
<p>A backend pipeline that provides functions to run a genome search on a given RNA model, parse the search results into a SQL database, and then filter the data for use in generating new RNA models. The pipeline is written in Python and utilizes the Infernal, HMMER, CMfinder, and R2R software packages. The pipeline is designed to be modular, so that it can be easily extended to incorporate new software packages and operations. The pipeline is also designed to be run on a remote server, so that it can be accessed by the frontend website. The pipeline is currently incomplete, but the following functions have been implemented:</p>
<ul>
<li><code>cmbuild_submit</code>: Starts Slurm job to build and calibrate a covariance model for a given RNA sequence alignment using Infernal.</li>
<li><code>cmsearch_submit</code>: Starts Slurm job to search a given RNA model against a given genome database using Infernal's CMsearch.</li>
<li><code>cmsearch_parse</code>: Parses the results of a search into a SQLite database.</li>
<li><code>cmsearch_analyze</code>: Analyzes the results of a search by plotting the E-value score distribution of the hits, filtering the hits by E-value score and uniqueness and reporting count numbers, and running R2R on the hits to visualize the resulting model.</li>
<li><code>runr2r</code>: Runs R2R on a given RNA model to visualize the results of a CMsearch search.</li>
<li><code>sto_reformat</code>: Reformats a given RNA sequence alignment from Stockholm to FASTA (unannotated) format.</li>
<li><code>run_cmfinder</code>: Starts a Slurm job to run CMfinder on a given FASTA sequence file to predict RNA secondary structure/folding.</li>
</ul>
<p>The pipeline is currently being redesigned to utilize a SQL database instead of a filesystem + in-memory data structures to store and manipulate data. The pipeline is also being redesigned to be more modular, so that it can be easily extended to incorporate new software packages and operations. However, this means that some functionality was reversed in the process of redesigning the pipeline, and some functionality is currently missing. The following functions are currently missing:</p>
<ul>
<li><code>cmsearch_compare</code>: Compares the results of two searches to determine the number of hits that are unique to each search, and the number of hits that are shared between them.</li>
</ul>
<p>In lieu of a frontend website, the pipeline is currently accessed through a command-line interface (CLI) that provides commands for running each of the pipeline functions, or the pipeline can be run directly from the Python interpreter. The CLI is implemented using the Python Click library. It is possible to start a process that runs the entire pipeline in the background on the cluster, and then check the status of the process later.</p>
<h3 id="2-genetic-algorithm">2. Genetic Algorithm</h3>
<p>A genetic algorithm that predicts RNA secondary structure/folding for a given RNA sequence alignment. The genetic algorithm is written in Python, and is meant to replace the CMfinder step of the pipeline, since CMfinder is out-of-date and uses the Perl programming language. The genetic algorithm is currently incomplete, but a prototype has been implemented, nick-named TrEAT, or Tree Evolution of Augmenting Topologies, which refers to the usage of a tree data structure to represent the context-free grammar of the RNA secondary structure/folding, and is strongly based on the famous <a href="https://nn.cs.utexas.edu/downloads/papers/stanley.ec02.pdf">Neuro-Evolution of Augmenting Topologies</a> (NEAT) algorithm for evolving neural networks. The following functionality has been implemented:</p>
<ul>
<li><code>Rule</code>: A class that represents a rule in the context-free grammar of the RNA secondary structure/folding. Each rule consists of a sequence that describes the primary structure of the RNA, with id's that refer to other rules in the grammar. Each rule should also have a probability of being chosen, but this functionality has not yet been implemented. The rule class also contains functions to mutate the rule, to generate a random rule, and to crossover two rules</li>
<li><code>Genome</code>: A class that represents a genome, which is a collection of 'rule' genes that encode a context-free grammer. The genome class also contains functions to mutate the genome, generate a random genome, and crossover two genomes. The crossover function is designed as in NEAT, using a historical marking scheme to align the genes of the two genomes to solve the competing conventions problem.</li>
<li><code>Organism</code>: A class that represents an organism, which is a genome that has a fitness score and belongs to a species.</li>
<li><code>Species</code>: A class that represents a species, which is a collection of organisms with similar genomes, and is used to maintain diversity in the population. Originally devised by Kenneth Stanley in the NEAT algorithm to solve the problem of protecting structural innovation in the population.</li>
<li><code>Population</code>: A class that represents a population, which is a collection of organisms and their species. The population class also contains functions to evolve the population, which is done by speciating the organisms, killing off the worst organisms in each species, and then breeding the remaining organisms in each species to generate new organisms.</li>
<li><code>Simulation</code>: A class that represents a simulation, which has a population and defines the operations of reproduction/evolution for the population. The simulation class also contains functions to run the simulation, which is done by running the reproduction/evolutionary operations for a given number of generations.</li>
</ul>
<p>Once completed, the genetic algorithm must be adapted to fit into the pipeline.</p>
<h3 id="3-frontend-website">3. Frontend Website</h3>
<p>A frontend website that provides an interface for RNA researchers to utilize the pipeline and its various operations independently, as well as a new RALEE-esque RNA alignment editor, to conduct experiments on sequence data without needing technical knowledge of the particular software packages. The frontend is incomplete, and is currently implemented using the Python Flask framework and HTML/CSS/JavaScript. Ideally, the frontend will be redesigned to utilize the React.js library. The following functionality has been implemented:</p>
<ul>
<li>User authentication and authorization using the Flask-Login library.</li>
<li>Admin password protection, since we plan to host the website on a public server.</li>
<li>A prototype of the RALEE-esque RNA alignment editor, which is currently implemented using a custom Stockholm file parser plus JavaScript to render the alignment. The prototype is incomplete, and does not allow for editing or saving the alignment. We plan to reprogram the Stockholm alignment parser to utilize a numpy array to represent the alignment, making column/row operations faster, and to utilize the React.js library to render the alignment for editing.</li>
</ul>
<p>Once the pipeline is completed, the frontend will be developed to enable users to 1) run pipeline operations step-by-step (a la carte?), and 2) run the entire pipeline, and to visualize the results of the operations. As of now, the frontend is not connected to the backend pipeline, and is only a prototype.</p>
<h2 id="methodology">Methodology</h2>
<p>Much deliberation was put into the design of the pipeline and the project as a whole. Some of the design decisions are described below – although these are subject to change as the project progresses.</p>
<h3 id="code-structure">Code Structure</h3>
<p>The code for the pipeline is organized into the following directories:</p>
<ul>
<li><code>data</code>: Contains the input and output data files for the pipeline, such as the RNA sequence alignments the CMs, R2R diagrams, and charts, separated into subdirectories for each step of the pipeline and for each RNA motif.</li>
<li><code>web</code>: Contains the code for the frontend website, separated into subdirectories for Flask blueprints, Jinja HTML templates, and JavaScript/CSS files.</li>
<li><code>treat</code>: Contains the code for the genetic algorithm.</li>
<li><code>cgk</code>: Contains some code files from Chris's project, which are currently unused but have been kept for reference. Methodology from these files has been incorporated into the pipeline.</li>
<li><code>jps</code>: Contains the pipeline code, separated into utility and file I/O functions, database models, pipeline operations, and bash scripts for running certain software packages.</li>
<li><code>cpsc290</code>: Contains the end-of-term report for the project – an abstract, a set of web pages describing the project, and a final project report in PDF format, which is a copy of this README file.</li>
</ul>
<h3 id="database-architecture">Database Architecture</h3>
<p>The pipeline uses a SQLite database to store the results of each step of the pipeline. The ORM (Object-Relational Mapping) library SQLAlchemy is used to interact with the database, which is organized into the following tables:</p>
<ul>
<li><code>SlurmJob</code>: Represents a Slurm job, including the job ID, arguments passed, and the job status. The SlurmJob table is used to keep track of the jobs that are running on the cluster, and to check their status.</li>
<li><code>Search</code>: Represents a CMsearch search, including the search parameters (cmfile, target database, and other command-line flags), a timestamp, the Slurm job ID associated with the search, and associated hits found by the search. The Search table is used to keep track of the CMsearch searches that have been run, and to store the results of the searches.</li>
<li><code>Hit</code>: Represents a hit found by a CMsearch search, including the search that found the hit, the target sequence/chromosome accession identifier, the start and end positions of the hit in the target sequence, the strand (3' to 5' or 5' to 3'), the E-value score of the hit, and the alignment of the hit, as well as a few other fields from the CMsearch output. The Hit table is used to keep track of and store the hits that have been found by CMsearch searches.</li>
<li><code>Alnseq</code>: Represents an alignment of a hit, including a reference to the multiple sequence alignment containing the hit and the actual text of the aligned sequence. The Alnseq table is used to keep track of and store the alignments of hits.</li>
<li><code>StoFeature</code>: Represents a feature of a Stockholm alignment, including a reference to the Alnseq, the feature format (GC, GF, GS, or GR), field (e.g. SS_cons, seq, or RF), and text. The StoFeature table is used to keep track of and store the features of Stockholm alignments so that important annotations can be persisted when hits are filtered and written to new Stockholm alignment files.</li>
</ul>
<h2 id="experimental-process-for-sodium-binding-riboswitches">Experimental Process for Sodium-Binding Riboswitches</h2>
<p>I am conducting an experiment using my pipeline to analyze two sodium-binding RNA motifs, known as nhaA-I and DUF1646, to determine whether they can be combined into a single representative model. My experimental process is as follows:</p>
<h3 id="1-download-motifs">1. Download motifs</h3>
<p>I downloaded the nhaA-I (RF03057) and DUF1646 (RF03071) RNA sequence alignments from the Rfam website. Shown below are R2R diagrams of the two alignments:</p>
<table>
<thead>
<tr>
<th></th>
<th>nhaA-I</th>
<th>DUF1646</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>R2R diagram</strong></td>
<td><img src="files/nha/RF03057.png" alt="nhaA-I R2R diagram"></td>
<td><img src="files/duf/RF03071.png" alt="DUF1646 R2R diagram"></td>
</tr>
</tbody>
</table>
<p>As shown by the diagrams, the two RNAs are very similar, each with two stem-loops. The two RNAs also have similar lengths, with nhaA-I being 95 nucleotides long, and DUF1646 being 77 nucleotides long.</p>
<h3 id="2-search-against-genome-database">2. Search against genome database</h3>
<p>Using my pipeline, I ran CMsearch on the two alignments against a representative sample of bacterial genomes from Genome Taxonomy Database Release 207. The analyses of the results of the CMsearch searches, filtered by uniqueness and E-value threshold, are shown below:</p>
<table>
<thead>
<tr>
<th></th>
<th>nhaA-I</th>
<th>DUF1646</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>Score distribution</strong></td>
<td><img src="files/nha/gtdb-prok_nhaA-I_2_score_distribution.png" alt="nhaA-I search score distribution"></td>
<td><img src="files/duf/gtdb-prok_DUF1646_1_score_distribution.png" alt="DUF1646 search score distribution"></td>
</tr>
<tr>
<td><strong>R2R diagram</strong></td>
<td><img src="files/nha/gtdb-prok_nhaA-I_2.uniq.keepE1.png" alt="nhaA-I search results R2R diagram"></td>
<td><img src="files/duf/gtdb-prok_DUF1646_1.uniq.keepE1.png" alt="DUF1646 search results R2R diagram"></td>
</tr>
<tr>
<td><strong># hits</strong></td>
<td>4365</td>
<td>1649</td>
</tr>
<tr>
<td><strong># unique hits</strong></td>
<td>4091</td>
<td>1478</td>
</tr>
<tr>
<td><strong># hits with E-value &lt; 1</strong></td>
<td>1796</td>
<td>497</td>
</tr>
<tr>
<td><strong># unique hits with E-value &lt; 1</strong></td>
<td>1663</td>
<td>421</td>
</tr>
</tbody>
</table>
<h3 id="3-refold-using-cmfinder">3. Refold using CMfinder</h3>
<p>I reformatted the search results alignments (Stockholm files) as FASTA sequence files, and ran CMfinder on each of 1) the nhaA-I search results alignment, 2) the DUF1646 search results alignment, and 3) the combined search results alignment generated by concatenating the two FASTA files. The CMfinder results are shown below. Chris created another combined alignment, dubbed LiNa, using <code>esl-alimerge</code> to combine the original nhaA-I and DUF1646 Rfam alignment files, also shown below.</p>
<table>
<thead>
<tr>
<th></th>
<th>nhaA-I</th>
<th>DUF1646</th>
<th>Combined</th>
<th>LiNa</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>R2R diagram</strong></td>
<td><img src="files/nha/nhaA-I.fna.motif.h2_1.png" alt="nhaA-I CMfinder results R2R diagram"></td>
<td><img src="files/duf/DUF1646.fna.motif.h2_1.png" alt="DUF1646 CMfinder results R2R diagram"></td>
<td><img src="files/combined/DUF1646_nhaA-I.fna.motif.h2_1.png" alt="combined CMfinder results R2R diagram"></td>
<td><img src="files/lina/lina-combo-v1.png" alt="LiNa R2R diagram"></td>
</tr>
</tbody>
</table>
<h3 id="4-search-new-models-against-genome-database">4. Search new models against genome database</h3>
<p>Then, to evaluate our combined models, I ran Infernal's <code>cmsearch</code> once more on the combined alignment generated by CMfinder, and on the LiNa alignment. The results of the searches are shown below:</p>
<table>
<thead>
<tr>
<th></th>
<th>Combined</th>
<th>LiNa</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>Score distribution</strong></td>
<td><img src="files/combined/DUF1646_nhaA-I.fna.motif.h2_1_gtdb-bact-r207-repr_E1000.0_score_distribution.png" alt="combined search score distribution"></td>
<td><img src="files/lina/lina-combo-v1_gtdb-bact-r207-repr_E1000.0_score_distribution.png" alt="LiNa search score distribution"></td>
</tr>
<tr>
<td><strong>R2R diagram</strong></td>
<td><img src="files/combined/DUF1646_nhaA-I.fna.motif.h2_1_gtdb-bact-r207-repr_E1000.0.uniq.keepE1.png" alt="combined search results R2R diagram"></td>
<td><img src="files/lina/lina-combo-v1_gtdb-bact-r207-repr_E1000.0.uniq.keepE100.png" alt="LiNa search results R2R diagram"></td>
</tr>
<tr>
<td><strong># hits</strong></td>
<td>14536</td>
<td>2354</td>
</tr>
<tr>
<td><strong># unique hits</strong></td>
<td>13398</td>
<td>2183</td>
</tr>
<tr>
<td><strong># hits with E-value &lt; 1</strong></td>
<td>1818</td>
<td>1953</td>
</tr>
<tr>
<td><strong># unique hits with E-value &lt; 1</strong></td>
<td>1665</td>
<td>1801</td>
</tr>
</tbody>
</table>
<p>As shown by the table, the combined model found far more hits than the LiNa model, but the LiNa model found more hits that were well-matched to the model (hits with E-value &lt; 1).</p>
<h3 id="5-evaluate-search-results">5. Evaluate search results</h3>
<p>Finally, I performed a few comparisons between the four searches, since we want to see how the combined models compare to the individual models.</p>
<p>First, we report hit counts for each search, including the total number of hits found, the number of unique hits found, and the number of unique hits found with an E-value less than 1000.0.</p>
<table>
<thead>
<tr>
<th>Search Name</th>
<th># Total</th>
<th># Unique</th>
<th># Unique E&lt;1000</th>
</tr>
</thead>
<tbody>
<tr>
<td>gtdb-prok_DUF1646_1</td>
<td>1649</td>
<td>1478</td>
<td>1474</td>
</tr>
<tr>
<td>gtdb-prok_nhaA-I_2</td>
<td>4365</td>
<td>4091</td>
<td>4080</td>
</tr>
<tr>
<td>DUF1646_nhaA-I.fna.motif.h2_1_gtdb-bact-r207-repr_E1000.0</td>
<td>14536</td>
<td>13398</td>
<td>13355</td>
</tr>
<tr>
<td>lina-combo-v1_gtdb-bact-r207-repr_E1000.0</td>
<td>2354</td>
<td>2183</td>
<td>2181</td>
</tr>
</tbody>
</table>
<p>Then, we check for intersecting hits between the original nhaA-I and DUF1646 searches, finding zero intersecting hits:</p>
<pre class="hljs"><code><div>Found 0 intersecting hits between DUF1646 and nhaA-I
</div></code></pre>
<p>Then, for each combined model, we compare the results of the combined model to the nhaA-I and DUF1646 searches to identify true positives (TP), false positives (FP), and false negatives (FN) by the following template:</p>
<table>
<tr><td>
<table>
<thead>
<tr>
<th>E</th>
<th>I</th>
<th>II</th>
<th>III</th>
<th>IV</th>
<th>V</th>
</tr>
</thead>
<tbody>
<tr>
<td>1</td>
<td>?</td>
<td>?</td>
<td>?</td>
<td>?</td>
<td>?</td>
</tr>
<tr>
<td>10</td>
<td>?</td>
<td>?</td>
<td>?</td>
<td>?</td>
<td>?</td>
</tr>
<tr>
<td>100</td>
<td>?</td>
<td>?</td>
<td>?</td>
<td>?</td>
<td>?</td>
</tr>
<tr>
<td>1000</td>
<td>?</td>
<td>?</td>
<td>?</td>
<td>?</td>
<td>?</td>
</tr>
</tbody>
</table>
</td><td>
</td><td>
<table>
<thead>
<tr>
<th>Column</th>
<th>Description</th>
<th>Meaning</th>
</tr>
</thead>
<tbody>
<tr>
<td>E</td>
<td>E-value threshold</td>
<td></td>
</tr>
<tr>
<td>I</td>
<td># of DUF1646 hits not in combo</td>
<td>FN</td>
</tr>
<tr>
<td>II</td>
<td># of DUF1646 hits in combo</td>
<td>TP</td>
</tr>
<tr>
<td>III</td>
<td># of combo hits not in DUF1646 or nhaA-I</td>
<td>FP*</td>
</tr>
<tr>
<td>IV</td>
<td># of nhaA-I hits in combo</td>
<td>TP</td>
</tr>
<tr>
<td>V</td>
<td># of nhaA-I hits not in combo</td>
<td>FN</td>
</tr>
</tbody>
</table>
<p>*Some of the false positives in III may actually be true positives. For example, if a hit is found in the combined model, but not in either of the individual models, it could be a true positive that was missed by the individual models, or it could be a false positive that was found by the combined model – we can't tell without further analysis.</p>
</td></tr>
</table>
<p>The results of these comparisons are shown below:</p>
<table>
<tr><td>
<p><strong>DUF1646_nhaA-I vs. DUF1646 and nhaA-I</strong></p>
<table>
<thead>
<tr>
<th>E</th>
<th>I</th>
<th>II</th>
<th>III</th>
<th>IV</th>
<th>V</th>
</tr>
</thead>
<tbody>
<tr>
<td>1000</td>
<td>421</td>
<td>0</td>
<td>11698</td>
<td>1657</td>
<td>6</td>
</tr>
<tr>
<td>100</td>
<td>421</td>
<td>0</td>
<td>2768</td>
<td>1653</td>
<td>10</td>
</tr>
<tr>
<td>10</td>
<td>421</td>
<td>0</td>
<td>577</td>
<td>1631</td>
<td>32</td>
</tr>
<tr>
<td>1</td>
<td>421</td>
<td>0</td>
<td>110</td>
<td>1555</td>
<td>108</td>
</tr>
</tbody>
</table>
</td><td>
<p><strong>LiNa vs. DUF1646 and nhaA-I</strong></p>
<table>
<thead>
<tr>
<th>E</th>
<th>I</th>
<th>II</th>
<th>III</th>
<th>IV</th>
<th>V</th>
</tr>
</thead>
<tbody>
<tr>
<td>1000</td>
<td>257</td>
<td>164</td>
<td>399</td>
<td>1618</td>
<td>45</td>
</tr>
<tr>
<td>100</td>
<td>278</td>
<td>143</td>
<td>82</td>
<td>1576</td>
<td>87</td>
</tr>
<tr>
<td>10</td>
<td>325</td>
<td>96</td>
<td>25</td>
<td>1388</td>
<td>275</td>
</tr>
<tr>
<td>1</td>
<td>395</td>
<td>26</td>
<td>14</td>
<td>990</td>
<td>673</td>
</tr>
</tbody>
</table>
</td></tr>
</table>
<p>From the table on the left we can see that my combined model overfits to nhaA-I and completely misses hits from DUF1646. This is likely due to an issue with CMfinder; since the nhaA-I search contributes more hits than DUF1646 (1663 &gt; 421) to the combined input to CMfinder, it is likely that CMfinder finds the pattern in nhaA-I and then stops looking for it in DUF1646. This is a problem that I will need to address in the future. However, we can see that the combined model does find some hits that are not found by either of the individual models, so it still may be useful. Future work will include further analysis of these hits, described in the next section.</p>
<p>From the table on the right we see that the LiNa combined model search also overfits to nhaA-I, but not as badly as the CMfinder combined model. This is likely due to the fact that LiNa uses a different algorithm to combine the models, and it is not as sensitive to the number of hits contributed by each model. The LiNa combined model finds a few hits that are not found by either of the individual models, but not as many as the CMfinder combined model.</p>
<h3 id="6-aside-report-tandem-hits">6. Aside: report tandem hits</h3>
<p>During model comparison, we also report any instances of tandems, which we define as hits that are located within 100 bp of each other. This information is not useful to our current analysis, but could be an interesting case study and demonstrates the pipeline's breadth of potential functionality. The results are shown below:</p>
<pre class="hljs"><code><div>Found tandem: MGTI01000074.1 11757-11703 11644-11589
Found tandem: MGTI01000074.1 11644-11589 11757-11703
Found tandem: DHQV01000099.1 4515-4574 4379-4429
Found tandem: MVRP01000073.1 1364-1414 1503-1556
Found tandem: MVRP01000073.1 1503-1556 1364-1414
Found tandem: JADJGV010000011.1 59728-59776 59636-59587
Found tandem: JADJGV010000011.1 59636-59587 59728-59776
Found tandem: VGRU01000033.1 489-431 354-296
Found tandem: VGRU01000033.1 354-296 489-431
Found tandem: MVRS01000109.1 9551-9494 9694-9633
Found tandem: DCVI01000207.1 44921-44868 44770-44713
Found tandem: DCVI01000207.1 44770-44713 44921-44868
Found tandem: VGSH01000014.1 30017-29962 30163-30108
Found tandem: VGSH01000014.1 30017-29962 29871-29813
Found tandem: VGSH01000014.1 30163-30108 30017-29962
Found tandem: VGSH01000014.1 29871-29813 30017-29962
Found tandem: MVRP01000001.1 47864-47916 47994-48052
Found tandem: MVRP01000001.1 47994-48052 47864-47916
Found tandem: MVRP01000044.1 25032-24978 25169-25119
Found tandem: MVRP01000044.1 25169-25119 25032-24978
Found tandem: MGTI01000074.1 11757-11703 11641-11590
Found tandem: MGTI01000074.1 11644-11589 11754-11704
Found tandem: MVRP01000073.1 1364-1414 1506-1555
Found tandem: MVRP01000073.1 1503-1556 1367-1414
Found tandem: JADJGV010000011.1 59728-59776 59633-59588
Found tandem: JADJGV010000011.1 59636-59587 59731-59775
Found tandem: VGRU01000033.1 489-431 351-297
Found tandem: VGRU01000033.1 354-296 486-432
Found tandem: QZJR01000105.1 17578-17640 17438-17496
Found tandem: DCVI01000207.1 44770-44713 44918-44868
Found tandem: VGSH01000014.1 30017-29962 30160-30109
Found tandem: VGSH01000014.1 30017-29962 29868-29814
Found tandem: VGSH01000014.1 30163-30108 30014-29963
Found tandem: VGSH01000014.1 29871-29813 30014-29963
Found tandem: MVRP01000001.1 47864-47916 47997-48051
Found tandem: MVRP01000001.1 47994-48052 47867-47915
Found tandem: MVRP01000044.1 25032-24978 25166-25119
Found tandem: MVRP01000044.1 25169-25119 25029-24979
</div></code></pre>
<h2 id="conclusion">Conclusion</h2>
<h3 id="discussion">Discussion</h3>
<p>Much time spent working on the project was devoted to learning the basics of bioinformatics and computational biology, as well as the various software packages used in the project. I learned how to use the Infernal, HMMER, CMfinder, and R2R software packages, as well as the command-line interface of the Yale HPC cluster and the Slurm job scheduler to run jobs on the cluster.</p>
<p>Many of our regular meetings were also spent discussing the design of the pipeline, backend database, and code organization. Since I am transitioning the pipeline to SQL, the backend database is currently incomplete, and the code organization is currently in flux. However, I have learned a lot about how to design a modular, extensible, and maintainable software project.</p>
<p>Over the semester of CPSC 290, I've made a great deal of progress on the project, but there is still much to be done before the pipeline is ready for use by the Breaker Lab. I plan to continue working on the project over the summer, by the end of which I hope to have a working prototype. I will continue my experiments on sodium-binding riboswitches with the pipeline over the summer as well, and I hope to have some interesting results to share when I'm done.</p>
<h3 id="future-research">Future Research</h3>
<p>I hope to create a platform to facilitate and democratize biological sequence-based research by automating key processes that currently require installing and running command-line tools and working with filesystems manually.</p>
<p>There are some steps to the pipeline that have not yet been implemented, but are planned for the future. These include:</p>
<ul>
<li><input type="checkbox" id="checkbox0"><label for="checkbox0">Completing the transition to SQL and implementing a new </label><code>cmsearch_compare</code> method to compare two searches using SQL queries to identify duplicate/overlapping hits and tandem repeats in the outputs of <code>infernal</code> searches.</li>
<li><input type="checkbox" id="checkbox1"><label for="checkbox1">A function to map the hits of a </label><code>cmsearch</code> search onto the phylogenetic tree contained in the Genome Taxonomy Database (GTDB) using R or Python to read the GTDB taxonomy tree and the Infernal output files, and then render a new file with the hits mapped onto the tree. This will enable us to view the distribution of hits across the tree and determine if there are any patterns in the distribution of hits that may be useful for identifying new riboswitches. For example, if a particular riboswitch is found in a particular clade of bacteria, it may be possible to use this information to identify new instances of the riboswitch in other, closely related bacteria. Or, if two riboswitches, such as nhaA-I and DUF1646, are found in the same clade of bacteria, this may indicate that the two riboswitches have common ancestry and can be combined into a single model. This is one method of determining whether a &quot;false positive&quot; hit from the combined search is actually a true positive that was missed by the original models, and therefore should be added to the model.</li>
<li><input type="checkbox" id="checkbox2"><label for="checkbox2">A function to access genetic context for a sequence of interest from the GTDB FASTA file and render a visualization of the genetic context of the sequence of interest. This will enable us to view gene associations of a given sequence, such as a cmsearch hit, and determine if the sequence is associated with a particular gene or gene cluster, which can tell us valuable information. For example, if a hit is upstream of a gene that is annotated as a transporter (e.g. a sodium ion transporter), it is likely that the sequence is a riboswitch that regulates the expression of that transporter gene. This is another method of determining whether a &quot;false positive&quot; hit from the combined search is actually a true positive that was missed by the original models, and therefore should be added to the model.</label></li>
</ul>
<p>In addition to these planned features, I plan to use the pipeline to update the Breaker Lab's collection of riboswitches. The Breaker Lab has a roster of ~50 riboswitches that have been identified in bacteria and eukaryotes, but have not been updated for a number of years. I am planning to utilize the pipeline to update the nhaA-I and DUF1646 motifs and other riboswitches using updated data, in order to identify more instances of these riboswitches and improve the models.</p>
<h3 id="acknowledgements">Acknowledgements</h3>
<p>I would like to thank my mentor, Chris, for his guidance, support, and instruction throughout the semester. I would also like to thank Dr. Ronald Breaker and the Breaker Lab for giving me the opportunity to work on this project.</p>

</body>
</html>
